src/common/test-jsonfds.c

```diff
@@ -0,0 +1,539 @@
+   * reset the structure for the next test.
+   */
+  gchar *buffer = helpers->inaccessible - (length + 1);
+  memcpy (buffer, value, length);
+  buffer[length] = '\0';
+
+  cockpit_json_print_string_property (helpers->stream, "key", buffer, -1);
+  helpers_compare_expected (helpers);
+  rewind (helpers->stream);
+
+  for (gint i = length; i < length + 3; i++)
+    {
+      cockpit_json_print_string_property (helpers->stream, "key", buffer, i);
+      helpers_compare_expected (helpers);
+      rewind (helpers->stream);
+    }
+}
+
+static void
+test_print_string (void)
+{
+  Helpers helpers;
+
+  helpers_init (&helpers);
+
+  /* create a string with every possible byte in it and check that
+   * everything is correctly escaped
+   */
+  char buffer[256];
+  int offset = 0;
+  char tmp[10];
+
+  helpers_expected (&helpers, ", \"key\": \"");
+
+  for (int c = 1; c < 32; c++) /* control characters (before space) */
+    {
+      /* ascii control characters printed as unicode escapes */
+      snprintf (tmp, sizeof tmp, "\\u%04x", c);
+      helpers_expected (&helpers, tmp);
+
+      buffer[offset++] = c;
+    }
+
+  helpers_expected (&helpers, " !");
+  buffer[offset++] = 32; /* space */
+  buffer[offset++] = 33; /* ! */
+
+  helpers_expected (&helpers, "\\\"");
+  buffer[offset++] = 34; /* " */
+
+  for (int c = 35; c < 92; c++) /* # through [ */
+    {
+      snprintf (tmp, sizeof tmp, "%c", c);
+      helpers_expected (&helpers, tmp);
+      buffer[offset++] = c;
+    }
+
+  helpers_expected (&helpers, "\\\\");
+  buffer[offset++] = 92; /* \ */
+
+  for (int c = 93; c < 127; c++) /* ] through ~ */
+    {
+      snprintf (tmp, sizeof tmp, "%c", c);
+      helpers_expected (&helpers, tmp);
+      buffer[offset++] = c;
+    }
+
+  helpers_expected (&helpers, "\\u007f");
+  buffer[offset++] = 127; /* DEL */
+
+  for (int c = 128; c < 256; c++) /* non-ascii */
+    {
+      helpers_expected (&helpers, "?");
+      buffer[offset++] = c;
+    }
+
+  buffer[offset++] = '\0';
+  g_assert_cmpint (offset, ==, sizeof buffer);
+  helpers_expected (&helpers, "\"");
+
+  /* print it with -1, correct length, and "too big" length */
+  cockpit_json_print_string_property (helpers.stream, "key", buffer, -1);
+  helpers_compare_expected (&helpers);
+  rewind (helpers.stream);
+
+  cockpit_json_print_string_property (helpers.stream, "key", buffer, 255);
+  helpers_compare_expected (&helpers);
+  rewind (helpers.stream);
+
+  cockpit_json_print_string_property (helpers.stream, "key", buffer, 256);
+  helpers_compare_expected (&helpers);
+  rewind (helpers.stream);
+
+  helpers_clear (&helpers);
+}
+
+static void
+test_print_string_memory_safety (void)
+{
+  gchar characters[] = { '\n', ' ', 'a', '\\', '\"', 0xcc };
+  const gchar *escaped[] = { "\\u000a", " ", "a", "\\\\", "\\\"", "?" };
+  gchar buffer[MAX_GROUPS * MAX_REPS];
+  Helpers helpers;
+
+  helpers_init (&helpers);
+
+  for (gint n_groups = MIN_GROUPS; n_groups <= MAX_GROUPS; n_groups++)
```

Comment:
In general I'm not a fan of making tests/test content generated like this. It makes it more difficult to read.

What is the benefit of doing it in a procedural manner versus writing a big json string into your expected_buffer and calling `test_string`? The latter would make it much more explicit, like in your `test_print_numeric`.

Reply:
This is testing thousands (or millions?) of possible combinations...

Reply:
can you at least explain what it's doing in a bit more detail then?

Reply:
done.